File: go/solutions/0011_container_with_most_water/container_with_most_water.go
package container_with_most_water

// Solution 1:
// Time: O(n), Space: O(1)
func maxArea(height []int) int {
	maxArea, left, right := 0, 0, len(height)-1
	for left < right {
		area := (right - left) * min(height[left], height[right]) // 以较矮的作为高计算面积
		maxArea = max(maxArea, area)
		// 只移动较矮的一边
		if height[left] < height[right] {
			left++
		} else {
			right--
		}
	}

	return maxArea
}

File: python/solutions/0011_container_with_most_water/container_with_most_water.py
class Solution:
    # Solution 1:
    # Time: O(n), Space: O(1)
    def maxArea(self, height: list[int]) -> int:
        max_area, left, right = 0, 0, len(height) - 1
        while left < right:
            area = (right - left) * min(height[left], height[right])
            max_area = max(max_area, area)
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1

        return max_area
